frontend/helpers/product-list-helpers.ts

```diff
@@ -109,3 +117,165 @@ export function getAdminLinks({
       },
    ];
 }
+
+export function updateProductListWithItemTypeOverrides(
+   productList: ProductList,
+   page: number,
+   itemType?: string
+): ProductList {
+   const productListCopy = { ...productList };
+   productListCopy.title = getTitle(productList, itemType);
+   productListCopy.metaTitle = getMetaTitle(productList, itemType);
+   productListCopy.description = getDescription(productList, page, itemType);
+   productListCopy.metaDescription = getMetaDescription(productList, itemType);
+   productListCopy.tagline = getTagline(productList, page, itemType);
+   return productListCopy;
+}
```

Comment:
>The copy gets references to array/object fields of the original due to the shallow destructure. If the copy's referential fields get modified, so do the original's. This can be unexpected and hard to debug.

Good point, although would you be fine with this solution if we deep copied instead? Because i don't think thats too hard

>This function encourages the caller to create a variable like productListModified. It's easy to accidentally use productList where you intend to use productListModified, and vice versa, since they are the same type.

Ok i think the intent was to move away from modifing data conditionally in various components and have all the necessary data in the productlist. @sterlinghirsh do you have a particular response to which way you want the data to be?

Personally i feel like this is segmented enough from the productList being used, but im open to others thoughts.

Reply:
@danielcliu-ifixit and I went back and forth on this design for a while. I think there are tradeoffs either way. First of all, I don't think it's necessary to use a deep copy here since we shouldn't be modifying these values anyway. Using a deep copy on the off chance that someone starts modifying values in the future seems wasteful. I am more sympathetic to the problem of having productList vs productListModified. One option is making the default value be "productListOriginal" or making the productListModified var in an outer component that passes it in as productList to the inner component to keep the naming consistent.

I'm not sure I understand the comment about defining the fallback logic near the callers, but I'm not attached to doing it in the product-list-helpers file if it's only used in one place. I would rather not put the fallback + template logic in each place the title, desc, etc are referenced. Centralizing that logic at a high level and simplifying the inner components makes more sense to me.

Reply:
You both make good points and changed my mind about certain things. I agree that a deep copy is overkill.

> I'm not sure I understand the comment about defining the fallback logic near the callers

Some of this data is only ever used in one place, e.g. the `metaTitle` is only used in `<MetaTitle>`. So I was suggesting that we could move the fallback logic from here to MetaTitle.

However, I agree that repeating fallback logic for fields used in multiple places isn't a good idea either, e.g. the `description` is used in `<HeroDescription>` and `<ProductListCard>`.

Another idea:
We could add item type specific data to the product model after we fetch it from the cache, since the model is stored in an item type agnostic way. That way, we could have access to the original data and the item type specific data in our UI. (e.g. we use the original product list `title` for related posts, but the item type `title` in the `<HeroSection>`). All the UI has to do is pick which one it's interested in (perhaps `productList.itemTypeOverrides.metaTitle ?? productList.metaTitle`)

```ts
type ProductListWithItemType = {
   itemTypeOverrides: {
      title: string,
      metaTitle: string,
      ...
   },
   ... // the rest of the base productlist fields
};
```

Or another idea:
We store item type override data fore every item type on the product list model (can be cached too). That way we don't need a new `ProductList` type either.

```ts
type ProductList = {
   itemTypeOverrides?: {
      Batteries: {
         title: string,
         metaTitle: string,
         ...
      },
      Screens: { ... },
      ...
   },
   ... // the rest of the base productlist fields
};
```

Then the UI can choose fallbacks like:
```ts
const itemType = useDevicePartsItemType();
productList.itemTypeOverrides?.[itemType]?.metaTitle ?? productList.metaTitle
```

Reply:
Storing it for every item type requires knowing every possible item type, which will be based on the Algolia result. Might be possible, but I don't really see the advantage of precalculating these values for the dozen item types the user isn't viewing.